### PR TITLE
Offset: `$.fn.offset()` on element inside shadowRoot

### PR DESCRIPTION
**Pull request by nazar-pc**
`$.fn.offset()` didn't worked on `html>...>shadowRoot>...>target_element`, while `target_element.getBoundingClientRect` worked fine with correct result.
The reason was that check for disconnected node made by checking whether `document` contains `target_element` fails, which is correct, element inside shadowDOM, not in main DOM.

That is why I've added recursive traversal from element to its parent until it reaches `document` or until `parent_element.parentNode` is `falsy` (for disconnected nodes it will be `null`, if pass some badly composed object - it may be `undefined`, so, just `falsy`, in this case we just return).
If we encounter `DocumentFragment` (`shadowRoot`), we take `parent_element.host` property instead of `parent_element.parentNode` in order to get parent element of that `shadowRoot`.
Also it was necessary to add explicit check for `target_element.getBoundingClientRect` property.

This is modern, Web Components-aware version of ea507b3e998126ae1f94f4fd1618994d645c9cc8

### CODE REVIEW

File: src/offset.js [hunk 1]
```diff
@@ -93,9 +94,26 @@ jQuery.fn.extend({
 
 		docElem = doc.documentElement;
 
-		// Make sure it's not a disconnected DOM node
-		if ( !jQuery.contains( docElem, elem ) ) {
-			return box;
+		// We have element in ShadowDOM, need recursive traversal from element to its parent
+		if ( elem.matches && elem.matches( ":host *" ) ) {
+			if ( !elem.getBoundingClientRect ) {
```
Comment (dmethvin) [inline src/offset.js]: Is there any situation where a browser would support shadowDOM but not `getBoundingClientRect`? seems unlikely.

**Comment by dmethvin** [comment]
This needs unit tests.

File: src/offset.js (hunk 1 above)
Comment (mzgol) [inline src/offset.js]: We don't check form `getBoundingClientRect` at all, I removed this `if` in my commit dropping older browsers.

Comment (dmethvin) [inline src/offset.js]: Oh right, that was for some old BlackBerry.

Comment (nazar-pc) [inline src/offset.js]: Removed that

**Comment by nazar-pc** [comment]
Just merged upstream and dropped `getBoundingClientRect` check.
@dmethvin, how to run unit test properly here? Because to test this we need shadowDOM support from browser.
I've tried `(document.body.shadowRoot === undefined ? skip : test)(...)`, but `skip` is unavailable.

**Comment by mzgol** [comment]
> skip is unavailable.

Try `QUnit.skip`; we'll need to migrate all those globals to the QUnit namespace some time, perhaps `skip` isn't exposed as a global at all since it's a new API.

**Comment by nazar-pc** [comment]
Thanks, @mzgol, it works.
Added unit test, will be skipped if browser do not have native support for shadowDOM.

**Comment by deshawnbw** [comment]
Anything holding this up? Encountered this issue while using web components as well.

**Comment by nazar-pc** [comment]
Please, read discussion #1784 and pull request #2043.
There is different pull request that addresses this issue, also code might be simplified for 3.0 due to dropping support for some old browsers.

**Comment by gibson042** [comment]
Closing in favor of #2043, which is just about ready to land.

**Pull request closed without merging**